scripts/release/branch_diff.py

```diff
@@ -11,18 +15,66 @@ def __init__(self, remote, name):
 
     @classmethod
     def parse(cls, raw):
+        raw = raw.strip()
         if " -> " in raw:
             return None
         else:
-            parts = raw.split('/')
-            return Branch(parts[0], parts[1])
+            split_point = raw.find('/')
+            if split_point:
+                return Branch(raw[:split_point], raw[(split_point + 1):])
+            else:
+                raise Exception("Unable to parse branch '{}'".format(raw))
+
+
+class Difference:
+    def __init__(self, past_version):
+        """We store the difference as a dictionary, where the keys are branches,
+        and each branch maps to a list of repos where that branch was found"""
+        self.diff = {}
+        self.add_main_diff(past_version)
+        for submodule in os.listdir("submodules"):
+            self.add_submodule_diff(submodule, past_version)
 
+    @property
+    def branches(self):
+        return self.diff.keys()
 
-def get_branches_at(revision):
-    raw = run("git branch -r --merged {}".format(revision))
-    parsed = [Branch.parse(b) for b in raw.split('\n')]
-    branches = set(b.name for b in parsed if b)
-    return branches
+    def add(self, branches, submodule):
+        for b in branches:
+            if b not in self.diff:
+                self.diff[b] = []
+            self.diff[b].append(submodule)
+
+    def add_main_diff(self, past_version):
+        self.add(Difference.get_branch_diff(past=past_version), "main")
+
+    def add_submodule_diff(self, submodule, past_main_repo_version):
+        full_path = join("submodules", submodule)
+        current_version = get_submodule_version(submodule)
+        previous_version = get_past_submodule_version(submodule,
+                                                      past_main_repo_version)
+        d = Difference.get_branch_diff(current_version, previous_version,
+                                       working_dir=full_path)
+        self.add(d, submodule)
+
+    @classmethod
+    def get_branches_at(cls, revision, working_dir=None):
+        raw = run("git branch -r --merged {}".format(revision),
```

Comment:
This seems like the only place where the `working_dir` option is used - perhaps simpler would be to use

```
run("git -C {} branch -r --merged {}".format(working_dir, revision))
```

since we use that directory changing approach elsewhere already (and then dropping that argument to run)

Reply:
I had a go at this, but it's a bit nasty. Because `working_dir` can be `None`, so then you have to conditionally include the `-C`. Whereas the current working dir argument to `subprocess.run` does the right thing when you pass it `None` for free.